### PR TITLE
[FLINK-21396][table] Resolve tables when calling Catalog.createTable/alterTable

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This ensures that the framework always resolves CatalogTable or CatalogView before calling `Catalog.createTable/alterTable`.  It is still possible to pass unresolved tables when calling the catalog directly but this depends on the catalog implementation. 

Note: The PR is still incomplete because we should also only work with resolved objects on the planner side.

## Brief change log

- Update `CatalogManager` to resolve tables and views
- Forward calls in `TableEnvironment` to `CatalogManager`
- Enable the interface introduced in #15098 

## Verifying this change

This change is already covered by existing tests, such as `CatalogTest` and many more.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogBaseTable.java [hunk 1]
```diff
@@ -18,18 +18,32 @@
...
+ * A common parent that describes the <i>unresolved</i> metadata of a table or view in a catalog.
+ *
+ * @see CatalogTable
+ * @see CatalogView
  */
+@PublicEvolving
 public interface CatalogBaseTable {
 
+    /** The kind of {@link CatalogBaseTable}. */
+    enum TableKind {
```
Comment (Airblader) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogBaseTable.java:39]: Is "TableKind" the correct abstraction to use if one value is "Table"? If we e.g. enumerate types of apples, then AppleKind.GRANNY_SMITH is reasonable, but AppleKind.APPLE sounds out of place. A "table" is not a "kind of table".

I sort of feel the same way here. `CatalogBaseTable` has that problem, too, but that class existed already, so we probably can't rename it easily now.

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java [hunk 2]
```diff
@@ -1095,7 +1095,7 @@ public void testTimestampNotNull() {
         TestCollectionTableFactory.initData(sourceData);
 
         tEnv().executeSql(
-                        "CREATE TABLE SourceTable(i INT, ts AS LOCALTIMESTAMP, WATERMARK FOR ts AS ts) "
+                        "CREATE TABLE SourceTable(i INT, ts AS CAST(LOCALTIMESTAMP AS TIMESTAMP(3)), WATERMARK FOR ts AS ts) "
```
Comment (Airblader) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1098]: Is this a drive-by change? Or was this necessary?

Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/FunctionITCase.java:1098]: It was necessary, because the new schema validator is a bit more strict when it comes to precision. Currently, we only support time attributes of precision 3. As documented here: https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/sql/create.html#watermark a couple of locations violated this.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogBaseTable.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/CatalogBaseTable.java:39]: We were also discussing about the name in the past but couldn't come up with a better name.

Some background: a view is commonly defined as a `virtual table`. See https://www.geeksforgeeks.org/sql-views/ or https://www.w3schools.com/sql/sql_ref_create_view.asp or Wikipedia. So we have to distinguish between a `real/base table` (= table) or `virtual table` (=view). The common name is still table unless you have a better idea :(

**Review by Airblader — APPROVED** [review]

**Comment by Aitozi** [comment]
Hi @twalthr. Sorry to ping you here. When I am doing some migration work for the hive connector from `TableSchema` to `ResolvedSchema
`. I have a question about the `Catalog`'s interface of  `createTable/alterTable` . Since the createTable and alterTable should always work with the resolved table. Why the interface does not enforce it? If not, the implementer can not use the resolved schema without a cast. What do you think of this?

**Comment by Aitozi** [comment]
Thanks @twalthr for your input. Because the `TableEnvironment#getCatalog` leak the internal `Catalog` to the user's face, it's not compatible to enforce the Catalog to interact with resolved table only, Right?

Recently, I'm trying to migrate `TableSchema` to new APIs in hive connector. And get some problems when dealing with the HiveCatalog. The best way I think is to let it deal with the ResolvedCatalogTable directly. Now, with your notes, I think it makes sense to do an extra cast internally.
